### ISSUE TITLE
Ncmesh works for hooke, doesnt work for general elasticity with matrix free

### ISSUE THREAD

**Issue by joetristano** [mfem/mfem#3858]
When looking at the hooke app, ncmeshes solve just fine with constrained dof on the hanging nodes.  When generalizing to linear elasticity by removing newton and adding form linear system, mfem blows up in debug mode when setting up the constrained operator. Stating that vectors arent the same size during a subtraction. This is basically a matrix free blend of example2 and hooke.

Interestingly, when the code is compiled in release mfem allows the solution to continue but the displacements are erroneous. 

When walking through the debugger in hooke, the constrained operator is never used.  How are the constraints of the hanging nodes set if the operator is never called?

Thanks in advance
Joe

**Comment by jtristano** [comment]
It is on the last line where the size of b and z are not equal for the NCMesh case after calling FormLinearSystem.   From the docs, I had assumed that FormLinearSystem will handle the constraints from the NCMesh

```
void ConstrainedOperator::EliminateRHS(const Vector &x, Vector &b) const
{
   w = 0.0;
   const int csz = constraint_list.Size();
   auto idx = constraint_list.Read();
   auto d_x = x.Read();
   // Use read+write access - we are modifying sub-vector of w
   auto d_w = w.ReadWrite();
   MFEM_FORALL(i, csz,
   {
      const int id = idx[i];
      d_w[id] = d_x[id];
   });
   // A.AddMult(w, b, -1.0); // if available to all Operators
   A->Mult(w, z);
   b -= z;
```

**Comment by vladotomov** [comment]
Hi @jtristano, can you provide some short code that does the setup and reproduces the issue?

**Comment by jandrej** [comment]
If Hooke works with NCMesh this is a happy accident. I never ran non-conforming meshes during development.

**Comment by jtristano** [comment]
@jandrej  What is necessary to allow ncmesh with matrix free on hexes?  We see other examples where ncmesh , AMR, and matrix free are working together.   Do the elasticity operators need to be extended somehow?

**Comment by vladotomov** [comment]
@jtristano 
- I'd like to run the reproducer, but the much of the code is missing. Try to provide a minimal reproducer.
- You said hooke+NC+PA works, does it work in `debug`?
- Does you ex2+hooke+NA+PA combo work with full assembly?

**Comment by joconnor819** [comment]
Howdy!
I have a slightly modified hooke.cpp here that will work if you replace the original miniapp hooke.cpp with it:
[hooke.zip](https://github.com/mfem/mfem/files/12587618/hooke.zip)

It lets us change between 
 * Conformal/Nonconformal meshes
 * Having a force/no force
 * Prescribed displacement/no prescribed displacement
 * Newton vs CG solve
Using the meshes here:
[meshes_for_hooke.zip](https://github.com/mfem/mfem/files/12587659/meshes_for_hooke.zip)

Using these 16 options, here are the results we get:
[Hooke_Results.zip](https://github.com/mfem/mfem/files/12587662/Hooke_Results.zip)

In short, it looks we are not applying forces correctly for the nc meshes. Prescribed displacements look fine.

I couldn't figure out an easy way to have this do full assembly. To my knowledge, the hooke miniapp is matrix-free.

We'd love any advice you can give on getting the forces to work using a nonconformal mesh.

Thank you,
Justin

**Comment by jtristano** [comment]
@tzanio , @jandrej  checking to see if the inputs we provided were sufficient? 

Thanks in advance!
Joe

**Comment by shani9383** [comment]
Hi,
I'm also trying to solve a simple elastic linear problem using the parallel hooke mini-app. I'm actually using a conforming mesh but encountering issues with applying traction (force) boundary conditions (instead of the prescribed displacement). when forming the parallel linear system I get issues with the size of B ( true DOF vector). it seems that B also includes the overlapping DOFs due to parallelism. 
I would appreciate any help on how to solve this issue. what am I missing?!
Thanks,
Shani
Main modification: 
VectorArrayCoefficient f(dimension);
   for (int i = 0; i < dimension-1; i++)
    {
       f.Set(i, new ConstantCoefficient(0.0));
    }
    {
       Vector pull_force(pmesh.bdr_attributes.Max());
       pull_force = 0.0;
       pull_force(1) = -1.0e-2;
       f.Set(dimension-1, new PWConstCoefficient(pull_force));
    }
 
   ParLinearForm *b = new ParLinearForm(&elasticity_op.h1_fes_);
   
   b->AddBoundaryIntegrator(new VectorBoundaryLFIntegrator(f));
   
   b->Assemble();
   ParGridFunction x(&elasticity_op.h1_fes_);
   x = 0.0;
   Operator *A;
   Vector B, X;
   elasticity_op.FormLinearSystem(ess_tdof_list, x, *b, A, X, B);
   std::cout << "linear system formed" << std::endl;
   std::cout << "B size: " << B.Size() << "   X size: " << X.Size() << std::endl;
   std::cout << "elasticity_op Height: " << elasticity_op.Height() << std::endl;
   std::cout << "A Height: " << A->Height() << std::endl;
   std::cout << "x size: " << x.Size() << "   ess_tdof_list size: " << 
   ess_tdof_list.Size() << std::endl;
   std::cout << "b size: " << b->Size() << std::endl;

**Comment by jtristano** [comment]
@vladotomov @jandrej  - any thing obvious jump out as to why hooke doesn't work with NCMeshes?   We're further along in our MFEM knowledge than we were when we posted this and as we look at the code, nothing jumps out at us as the restrictions and prolongations should be adequate during the constrainedoperator instantiation, correct?
